flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/statistics/ContinuousHistogram.scala

```diff
@@ -0,0 +1,343 @@
+      */
+    for (i <- 1 to bins - 1) {
+      val tmpSum = count(getValue(i))
+      if (currSum <= wantedSum && wantedSum < tmpSum) {
+        val neededSum = wantedSum - currSum
+        val a: Double = getCounter(i) - getCounter(i - 1)
+        val b: Double = 2 * getCounter(i - 1)
+        val c: Double = -2 * neededSum
+        if (a == 0) {
+          return getValue(i - 1) + (getValue(i) - getValue(i - 1)) * (-c / b)
+        } else {
+          return getValue(i - 1) +
+            (getValue(i) - getValue(i - 1)) * (-b + Math.sqrt(b * b - 4 * a * c)) / (2 * a)
+        }
+      } else {
+        currSum = tmpSum
+      }
+    }
+    require(upper < MaxValue, "Set an upper bound before proceeding")
+    // this means wantedSum > sum(getValue(bins-1))
+    // this will likely fail to return a bounded value.
+    // Make sure you set some proper limits on min and max.
+    getValue(bins - 1) + Math.sqrt(
+      Math.pow(upper - getValue(bins - 1), 2) * 2 * (wantedSum - currSum) / getCounter(bins - 1))
+  }
+
+  /** Returns the probability (and by extension, the number of points) for the value b
+    * Since this is a continuous histogram, return the cumulative probability at b
+    *
+    * @return Cumulative number of points at b
+    */
+  def count(b: Double): Int = {
+    require(bins > 0, "Histogram is empty")
+    if (b < lower) {
+      return 0
+    }
+    if (b >= upper) {
+      var ret = 0
+      for (i <- 0 to bins - 1) {
+        ret = ret + getCounter(i)
+      }
+      return ret
+    }
+    /** Suppose x is the index with value just less than or equal to b
+      * Then, sum everything up to x-1
+      * Add half the value at x
+      * Find area of the trapezoid for x and the value b
+      */
+
+    val index = find(b) - 1
+    var m_b, s: Double = 0
+    if (index == -1) {
+      m_b = getCounter(index + 1) * (b - lower) / (getValue(index + 1) - lower)
+      s = m_b * (b - lower) / (2 * (getValue(index + 1) - lower))
+      return s.round.toInt
+    } else if (index == bins - 1) {
+      m_b = getCounter(index) +
+        (-getCounter(index)) * (b - getValue(index)) / (upper - getValue(index))
+      s = (getCounter(index) + m_b) *
+        (b - getValue(index)) / (2 * (upper - getValue(index)))
+    } else {
+      m_b = getCounter(index) + (getCounter(index + 1) - getCounter(index)) *
+        (b - getValue(index)) / (getValue(index + 1) - getValue(index))
+      s = (getCounter(index) + m_b) *
+        (b - getValue(index)) / (2 * (getValue(index + 1) - getValue(index)))
+    }
+    for (i <- 0 to index - 1) {
+      s = s + getCounter(i)
+    }
+    s = s + getCounter(index) / 2
+    s.round.toInt
+  }
+
+  /** Returns the total number of entries in the histogram
+    *
+    * @return Total number of points added in the histogram
+    */
+  override def total: Int = {
+    var result = 0
+    for (i <- 0 to bins - 1) {
+      result = result + getCounter(i)
+    }
+    result
+  }
```

Comment:
`total` method can be replaced by following. `getCounter` method checks range. But in this case, we don't need to check range.

```scala
override def total: Int = data.iterator.map(_._2).sum
```